### ISSUE TITLE
After an N amount of time, putObject payloads become UNSIGNED

### ISSUE THREAD

**Issue by hellsan631** [aws/aws-sdk-js#1063]
__Error:__ SignatureDoesNotMatch, 
__Message:__ The request signature we calculated does not match the signature you provided. Check your key and signing method.
__Node Version:__ 0.12.x (we deploy to heroku using CircleCI)

This is a bug we've been trying to fix for about 2 weeks now. Here is our code

```js
var AWS     = require('aws-sdk'); //v2.4.7
var multer  = require('multer');  //v1.1.0
var express = require('express'); //v4.2.0
var upload  = multer();
var router  = express.Router();

AWS.config.loadFromPath(__dirname + '/../../aws.config.json');

router.post('/upload/image', upload.single('file'), function (req, res, next) {
  var file = res.file;
  var bucket = new AWS.S3({
    signatureVersion: 'v4',
    params: {
      Bucket: 'images.magnises.com'
    }
  });

  var params = {
    ACL:'public-read',
	Key: file.name,
	ContentType: file.type,
	ContentEncoding: 'base64',
	Body: file.buffer,
	ServerSideEncryption: 'AES256'
  };

  bucket.putObject(params, function (err, data) {
    if (err) return next(err);

    res.status(200).send({ message: 'Image successfully uploaded' });
  });
});
```

The bug happens after an N amount of time on our _production_ only, meaning its most likely tied to some memory issues realting to ~~user activity~~. It's usually after about an hour or two or three of solid use.

Here are things that _do not_ trigger the bug immediately.
- ~~Uploading Event Images from the admin~~
- ~~Uploading Profile Images on Login UserInfo Page~~
- ~~Uploading Profile  Images on User Profile Page~~
- ~~Uploading Company Images from admin~~

Generally any usage of that API does not trigger the bug.

__Broken Request__
```js
{
  code: "SignatureDoesNotMatch",
  extendedRequestId: "Ic5zxY4r+esmZt+vh32iawnEKFfjsvNpgxX3Uj9vZjac786G/pg44xNEvFV/BEfUgh+iSWrSI9I=", 
  message: "The request signature we calculated does not match the signature you provided. Check your key and signing method.", 
  region: "None", 
  request: {
    body: {...}, 
    endpoint: {
      host: "s3.amazonaws.com", 
      hostname: "s3.amazonaws.com", 
      href: "https://s3.amazonaws.com/", 
      path: "/", 
      pathname: "/", 
      port: "443", 
      protocol: "https:"
    }, 
    headers: {
      Authorization: "AWS4-HMAC-SHA256 Credential=our_key/20160715/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=25c9389845c385b545db8773bc50881ba5f5329fa134af0874d21c9example", 
      Content-Length: 95, 
      Content-MD5: "caUNu6RMeBKLIht997tR8Q==", 
      Content-Type: "application/octet-stream", 
      Host: "s3.amazonaws.com", 
      User-Agent: "aws-sdk-nodejs/2.4.7 linux/v0.12.15", 
      X-Amz-Content-Sha256: "UNSIGNED-PAYLOAD", 
      X-Amz-Date: "20160715T195635Z"
    }, 
    method: PUT, 
    path: "/images.magnises.com/profile/56687997d1ad6c2b0037dfe5", 
    region: "us-east-1", 
    stream: {...}
  }, 
  requestId: "81D5FE98DC20C21B", 
  response: {
    body: {...}
  }, 
  retryable: false, 
  retryDelay: 66.41066682059318, 
  time: "2016-07-15T19:56:35.406Z"
}
```

__Working Request__
```js
{
  request: {
    body: {...}, 
    endpoint: {
      host: "s3.amazonaws.com", 
      hostname: "s3.amazonaws.com", 
      href: "https://s3.amazonaws.com/", 
      path: "/", 
      pathname: "/", 
      port: "443", 
      protocol: "https:"
    }, 
    headers: {
      Authorization: "AWS4-HMAC-SHA256 Credential=our_key/20160716/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=19ca3429cfc8852e7ed81138ab18708157e92fc5de24cba06a13ec2cexample", 
      Content-Length: 18735, 
      Content-MD5: "caUNu6RMeBKLIht997tR8Q==", 
      Content-Type: "application/octet-stream", 
      Host: "s3.amazonaws.com", 
      User-Agent: "aws-sdk-nodejs/2.4.7 linux/v0.12.15", 
      X-Amz-Content-Sha256: "384ee477c19d1177f0dfcb3d89e2231d5aaba3b8df4a123e07e66bb6example", 
      X-Amz-Date: "20160715T195635Z"
    }, 
    method: PUT, 
    path: "/images.magnises.com/profile/56687997d1ad6c2b0037dfe5", 
    region: "us-east-1", 
    stream: {...}
  }, 
  requestId: "E889099E639FF82B", 
  response: {
    body: {...}
  }, 
  retryable: false,
}
```

So for some reason, we are no longer signing our payloads, as "X-Amz-Content-Sha256" has changed to "UNSIGNED-PAYLOAD".

Here are things we've tried: 
- Hard-coding Keys
- Loading Keys through Config
- Environmental Keys
- Rebooting (works temporarily, until such time as the problem appears again)
- Reproducing on dev (heroku, configured to be a 1:1 with production, _unable to reproduce_)
- Using the same AWS version in Node and Browser (we use one in the front-end admin area for GET requests, one was an older version)
- Enabling request retries
- Changing signature version to v2 (errors with "timestamp undefined", no longer works)

The only other active use of the AWS-SDK in our project is when users get their profile photo, which is  ```bucket.headObject```, but this happens right after an image upload, so I doubt it is the cause.

**Comment by chrisradek** [comment]
@hellsan631 
We'll need to investigate further why your application is working for N amount of time, but one thing you can try is to set `s3DisableBodySigning: false` on your S3 client. I'm not sure yet why your requests are signed for N time, then unsigned, but can you try setting that parameter to make sure your payloads are always signed? (Assuming that's what you want).

**Comment by hellsan631** [comment]
Just an update. I've tried several things and it looks like it's an AWS caching issue.

Setting ```s3DisableBodySigning: false``` makes all things signed, but the requests still fail after a certain amount of time. I also tried ```correctClockSkew: true``` thinking that the date had something to do with it, but unfortunately that didn't work either.

That lead me to believe the payload was being manipulated while waiting for an async call. We are sending the payload via a buffer, so I converted the buffer to a base64 string, created an immutable copy of the variable, but still the error __SignatureDoesNotMatch__ persisted.

The goal of the function is to just upload a profile picture from an angular front-end (using [this](https://github.com/danialfarid/ng-file-upload)), in which I thought I was using some pretty boilerplate code.

Basically, a user upload an image which is storied in-memory as a buffer, then the image is sent to [tinify](https://github.com/tinify/tinify-nodejs), resized, and compressed, then sent back to us and converted to a new buffer. This new buffer would be uploaded to AWS. 

Lucky for us, tinify has AWS S3 built in (not using the SDK, but creating its own API call), which we are using now to upload to AWS. This works just fine at the moment. If you have any more suggestions or need any more information, I would be glad to help in any way that I can to solve this.
